core/desugarDatatypes.ml

```diff
@@ -180,45 +185,69 @@ struct
         | Table (r, w, n) -> `Table (datatype var_env r, datatype var_env w, datatype var_env n)
         | List k -> `Application (Types.list, [`Type (datatype var_env k)])
         | TypeApplication (tycon, ts) ->
+            let exception Kind_mismatch (* TODO add more information *) in
+            (* Matches kinds of the quantifiers against the type arguments.
+             * Returns Types.type_args based on the given frontend type arguments. *)
+            let match_quantifiers qs =
+              let match_kinds (q, t) =
+                let primary_kind_of_type_arg : Datatype.type_arg -> PrimaryKind.t = function
+                  | Type _ -> PrimaryKind.Type
+                  | Row _ -> PrimaryKind.Row
+                  | Presence _ -> PrimaryKind.Presence
+                in
+                if primary_kind_of_quantifier q <> primary_kind_of_type_arg t then
+                  raise Kind_mismatch
+                else (q, t)
+              in
+              let type_arg' var_env alias_env = function
+                | Row r -> `Row (effect_row var_env alias_env r)
+                | t -> type_arg var_env alias_env t
+              in
+              begin
+                try
+                  let ts = ListUtils.zip' qs ts in
+                  List.map
+                    (fun (q,t) ->
+                      let (q, t) = match_kinds (q, t) in
+                      match subkind_of_quantifier q with
+                      | (_, Restriction.Effect) -> type_arg' var_env alias_env t
+                      | _ -> type_arg var_env alias_env t) ts
+                with
+                | ListUtils.Lists_length_mismatch ->
+                   failwith (Printf.sprintf
```

Comment:
In line with what was said in #525, would it be possible to replace `failWith` with `Errors`?